### PR TITLE
acme service: letsencrypt cert auto-renewal with simp_le

### PR DESCRIPTION
**Pull request by fpletz**
This new service invokes `simp_le` for a defined set of certs on a regular basis with a systemd timer. `simp_le` is smart enough to handle account registration, domain validation and renewal on its own. The only thing required is an existing HTTP server that serves the path `/.well-known/acme-challenge` from the webroot cert parameter.
### Example

```
  security.acme.certs."foo.example.com" = {
    webroot = "/var/www/challenges";
    extraDomains = { "www.example.com" = null; };
    email = "foo@example.com";
    validMin = 2592000;
    renewInterval = "weekly";
  };
```
### Example Nginx vhost

```
  services.nginx.appendConfig = ''
    http {
      server {
        server_name _;
        listen 80;
        listen [::]:80;

        location /.well-known/acme-challenge {
          root /var/www/challenges;
        }

        location / {
          return 301 https://$host$request_uri;
        }
      }
    }
  '';
```
### Open TODOs
- [x] Notify/reload/restart dependent services
- [x] ~~Provide a basic HTTP Server for verification if host doesn't provide one (e.g. python SimpleHTTPServer)~~
- [x] Don't run `simp_le` as root and use cert-specific owner/group/mode at least for private keys

Feedback and suggestions for improvements would be much appreciated.

### CODE REVIEW

File: nixos/modules/services/security/simp_le.nix [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        after = [ "network.target" ];
+        serviceConfig = {
+          Type = "oneshot";
+          SuccessExitStatus = "0 1";
+        };
+        preStart = ''
+          mkdir -p "${cfg.directory}/${cert}"
+        '';
+        script = ''
+          WEBROOT="${optionalString (data.webroot == null) data.webroot}"
```
Comment (lethalman) [inline nixos/modules/services/security/simp_le.nix]: `WEBROOT` will be always empty with that null check...

**Comment by abbradar** [comment]
Do we really want separate `renewInterval`s and `validMin`s for `certOpts`? I can imagine no use case for this.

**Comment by abbradar** [comment]
I have several patches on top of your branch at https://github.com/abbradar/nixpkgs/commits/simp_le-new -- what do you think?

**Comment by globin** [comment]
@abbradar I'd always want to renew a certain time before the certificate expires.

**Comment by abbradar** [comment]
Sorry, bad wording. What I meant is: do we want those settings to be different for every domain?

**Comment by edolstra** [comment]
The option name `services.simp_le.certs` seems ugly to me. Isn't the use of `simp_le` just an implementation detail? Might be nicer to call it something like `security.letsEncrypt.certs`.

**Comment by jgillich** [comment]
@edolstra Let's Encrypt is also just a [ACME](https://en.wikipedia.org/wiki/Automated_Certificate_Management_Environment) provider, so we should probably just call it `acme`. Would surely avoid confusion down the road when (if) other CAs support the protocol.

**Comment by jeaye** [comment]
+1 Ah, I'd love to see this merged in.

**Comment by edolstra** [comment]
@jgillich Yeah, sounds good.

**Comment by domenkozar** [comment]
Good job :+1: Let's fix small things based on feedback and merge this.

**Comment by fpletz** [comment]
@abbradar Sorry for not getting back to you sooner, will look at this later today and incorporate your changes. Thanks! :beers:

**Comment by abbradar** [comment]
One of remaining things that bothers me is my own usage of `sudo` which is affected by `/etc/sudoers`. Is this a bad thing (e.g., can someone make such a config (perhaps invalid one) that `sudo` will reject dropping `root` privileges)? Would `su` be better?

**Comment by fpletz** [comment]
@abbradar Thanks for your patches. I've rebased and updated this PR with those.

Other changes:
 - `sudo` was removed and systemd will set the user and the group instead. A little hackery was needed to run the `postRun` hook as root in the `postStop` systemd hook only if the exit code is 0.
 - `services.simp_le` was renamed to `security.acme` as suggested by @edolstra & @jgillich.

**Comment by abbradar** [comment]
@fpletz I quite like your hack, nice!

One little thing (see my comment above): do we want `renewIntervals` and `validMins` to be different for every domain? I can't imagine any use case for this, and to simplify changing them for user we may want to move them to the top-level `security.acme`.

File: nixos/modules/security/acme.nix [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+          <filename>.well-known/acme-challenge/</filename> directory
+          will be created automatically if it doesn't exist.
+          <literal>http://example.org/.well-known/acme-challenge/</literal> must also
+          be available (notice unencrypted HTTP).
+        '';
+      };
+
+      validMin = mkOption {
+        type = types.int;
+        default = 2592000;
```
Comment (domenkozar) [inline nixos/modules/security/acme.nix]: for easier understanding I'd do `30 * 24 * 3600`

**Comment by domenkozar** [comment]
Would be extremely nice to add `meta.doc` attribute to the module with a minimal example for nginx.

**Comment by fpletz** [comment]
@abbradar You're right about `renewInterval` and `validMin`. Fixed.

@domenkozar Excellent idea. Added. :smile:

**Comment by abbradar** [comment]
Thanks! No more complaints from me, I think this is ready. BTW I haven't thought of  (or seen) adding `meta.maintainers` to a NixOS module this way -- very nice, I'll steal this idea ~_^.

**Comment by domenkozar** [comment]
Awesome work. As soon as someone confirms it works I'm +1 for merging (sorry, a bit busy now but I'll use this asap)

**Comment by abbradar** [comment]
Pushed with a small fix in https://github.com/NixOS/nixpkgs/commit/52505823969c9f7462fbb7efd5f36e32ca92b581. Works for me both for initial certificate and for checking (when `postRun` hook is skipped). Thank you for the work, @fpletz!

**Comment by jeaye** [comment]
Great work, folks! Stoked to see this one come in.

**Comment by fpletz** [comment]
Thanks for all the feedback! We have some more stuff coming up to improve the nginx module and optional automatic ACME support. :)

**Comment by jeaye** [comment]
On Sun, Dec 13, 2015 at 06:42:53AM -0800, Franz Pletz wrote:
>    Thanks for all the feedback! We have some more stuff coming up to improve
>    the nginx module and optional automatic ACME support. :)

Why's nginx getting all the love? What about the apache users? :D

**Comment by fpletz** [comment]
Well, if you look at the apache module and compare it to nginx, apache got all the love in the past. :) But we will probably hook in Apache and the other supported HTTP servers, too, if their respective modules already have support for vhosts.

**Comment by jeaye** [comment]
Aside from being on unstable, what's needed in order to use this now? A simple config lands me with

```text
error: The option `security.acme' defined in `/etc/nixos/service/acme.nix' does not exist.
(use ‘--show-trace’ to show detailed location information)
```

Config:

```nix
  security.acme =
  {
    certs =
    {
      "pastespace.org" =
      {
        webroot = "/home/http/acme.pastespace.org";
        extraDomains =
        {
          "www.pastespace.org" = null;
          "mail.pastespace.org" = null;
        };
        email = "contact@jeaye.com";
        user = "acme";
        group = "acme";
        postRun = "systemctl reload httpd.service";
      };
    };
  };
```

**Comment by jgillich** [comment]
It is in master, I dont think it's in unstable already.


On Mon, Dec 14, 2015, at 08:55 AM, jeaye wrote:
> Aside from being on unstable, what's needed in order to use this now?
> A simple config lands me with


> error: The option `security.acme' defined in
> `/etc/nixos/service/acme.nix' does not exist.
(use ‘--show-trace’ to show detailed location information)
>
> Config:


> security.acme =
  {  certs =  {    "pastespace.org" =    {      webroot =
  "/home/http/acme.pastespace.org";      extraDomains =      {
  "www.pastespace.org" = null;        "mail.pastespace.org" = null;
  };      email = "contact@jeaye.com";      user = "acme";      group =
  "acme";      postRun = "systemctl reload httpd.service";    };  }; };
> — Reply to this email directly or view it on GitHub[1].


>


Links:

  1. https://github.com/NixOS/nixpkgs/pull/11506#issuecomment-164371664

**Pull request merged**